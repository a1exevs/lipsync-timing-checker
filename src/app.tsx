import React, { ChangeEvent, useRef, useState } from 'react';
import WavesurferPlayer from '@wavesurfer/react';

const App: React.FC = () => {
  const [audioUrl, setAudioUrl] = useState<string>('');
  const [wavesurfer, setWavesurfer] = useState<any>(null);
  const [isPlaying, setIsPlaying] = useState(false);
  const jsonDataContainerRef = useRef<HTMLDivElement>(null);

  const onReady = (ws: any) => {
    setWavesurfer(ws);
    setIsPlaying(false);
    updateCurrentTime();
  };

  const onPlayPause = () => {
    wavesurfer && (wavesurfer as any).playPause();
  };

  const onFileChange = (event: ChangeEvent<HTMLInputElement>): void => {
    debugger;
    if ((event as any).target.files.length > 0) {
      const file = (event as any).target.files[0];
      const url = URL.createObjectURL(file);
      setAudioUrl(url);
    }
  };

  const onJSONFileChange = (event: ChangeEvent<HTMLInputElement>): void => {
    debugger;
    if ((event as any).target.files.length > 0) {
      const file = (event as any).target.files[0];
      const reader = new FileReader();
      reader.onload = function (e) {
        try {
          debugger;
          const jsonData = JSON.parse((e as any).target.result);
          displayJsonData(jsonData);
        } catch (error) {
          alert('Некорректный JSON файл');
        }
      };
      reader.readAsText(file);
    }
  };

  const displayJsonData = (jsonData: any) => {
    // Сброс содержимого, кроме заголовка
    const jsonDataContainer: any = jsonDataContainerRef.current;
    jsonDataContainer.innerHTML = '<h3>Слова и фонемы</h3>';

    // Создаем ползунок для слов и фонем
    const indicator = document.createElement('div');
    indicator.id = 'json-data-indicator';
    indicator.className = 'current-time';
    jsonDataContainer.appendChild(indicator);

    const duration = wavesurfer.getDuration();

    jsonData.words.forEach((wordItem: any) => {
      // Создание блока для слова
      const wordElement = document.createElement('div');
      wordElement.className = 'word';
      wordElement.textContent = wordItem.word;

      // Вычисляем позицию и ширину слова
      const wordStartRatio = wordItem.start / duration;
      const wordEndRatio = wordItem.end / duration;
      const wordLeft = wordStartRatio * 100;
      const wordWidth = (wordEndRatio - wordStartRatio) * 100;
      wordElement.style.left = `${wordLeft}%`;
      wordElement.style.width = `${wordWidth}%`;

      jsonDataContainer.appendChild(wordElement);

      // Создание блока для фонем
      if (wordItem.phonemes && Array.isArray(wordItem.phonemes)) {
        wordItem.phonemes.forEach((phonemeItem: any) => {
          const phonemeElement = document.createElement('div');
          phonemeElement.className = 'phoneme';
          phonemeElement.textContent = phonemeItem.phoneme;

          // Вычисляем позицию и ширину фонемы
          const phonemeStartRatio = phonemeItem.start / duration;
          const phonemeEndRatio = phonemeItem.end / duration;
          const phonemeLeft = phonemeStartRatio * 100;
          const phonemeWidth = (phonemeEndRatio - phonemeStartRatio) * 100;
          phonemeElement.style.left = `${phonemeLeft}%`;
          phonemeElement.style.width = `${phonemeWidth}%`;

          // Смещаем фонемы ниже слов
          phonemeElement.style.top = '30px'; // Отступ от слов
          jsonDataContainer.appendChild(phonemeElement);
        });
      }
    });
  };

  const updateCurrentTime = (): void => {
    if (!wavesurfer) {
      return;
    }
    const duration = wavesurfer.getDuration();
    const currentTime = wavesurfer.getCurrentTime();
    const percent = (currentTime / duration) * 100;

    // Обновляем позицию ползунка для слов и фонем
    const jsonDataIndicator = document.getElementById('json-data-indicator');
    if (jsonDataIndicator) {
      jsonDataIndicator.style.left = `${percent}%`;
    }
  };

  return (
    <>
      <h1>Audio Waveform and JSON Viewer</h1>
      <div id="waveform">
        {audioUrl && (
          <WavesurferPlayer
            height={200}
            waveColor="violet"
            url={audioUrl}
            onReady={onReady}
            onAudioprocess={updateCurrentTime}
            onSeeking={updateCurrentTime}
            onPlay={() => setIsPlaying(true)}
            onPause={() => setIsPlaying(false)}
          />
        )}
      </div>
      <button onClick={onPlayPause}>{isPlaying ? 'Pause' : 'Play'}</button>
      <div className="controls">
        <label htmlFor="audio-upload">Выбрать аудио:</label>
        <input type="file" id="audio-upload" accept="audio/*" onChange={onFileChange} />
        <label htmlFor="json-upload">Выбрать JSON:</label>
        <input type="file" id="json-upload" accept="application/json" onChange={onJSONFileChange} />
      </div>
      <div className="json-data" ref={jsonDataContainerRef}>
        <h3>Слова и фонемы:</h3>
        <div className="current-time" id="current-time"></div>
      </div>
    </>
  );
};

export default App;
