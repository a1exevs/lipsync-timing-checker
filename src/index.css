@import url('https://fonts.googleapis.com/css2?family=Roboto:ital,wght@0,100;0,300;0,400;0,500;0,700;0,900;1,100;1,300;1,400;1,500;1,700;1,900&display=swap');

* {
  padding: 0;
  margin: 0;
  box-sizing: border-box;
  font-family: 'Roboto', sans-serif;
}

* {
  scrollbar-color: #505354 #707374;
  scrollbar-width: thin;
}

body {
  margin: 0 auto;
  font-family: 'Roboto', sans-serif;
  background: #000000;

  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5rem;
  color: #e8e6e3;
}

.App {
  background: rgba(24, 26, 27, 0.75);
  margin: 0 auto;
  max-width: 1920px;
  min-height: 100vh;
  padding: 0;
}

.App__Title {
  text-align: center;
}

#waveform {
  display: flex;
  height: 200px;
  overflow-x: auto;
  overflow-y: hidden;
  border: 1px solid #ccc;
}

.current-time {
  position: absolute;
  top: 0;
  height: 100%;
  width: 2px;
  background-color: red;
  pointer-events: none;
}

.Controls {
  display: flex;
  align-items: center;
  gap: 16px;
  padding: 8px 16px;
}

.json-data-container {
  position: relative;
  height: 100px;
  border: 1px solid #ccc;
  overflow-x: auto;
  overflow-y: hidden;
}

.json-data {
  position: relative;
  height: 100px;
}

#json-data-indicator {
  position: absolute;
  top: 0;
  height: 100%;
  width: 2px;
  background-color: red;
  pointer-events: none;
  z-index: 10;
}

.word,
.phoneme {
  color: black;
  position: absolute;
  padding: 5px 10px;
  border: 1px solid #000;
  border-radius: 4px;
  cursor: pointer;
  transform: translateY(-50%);
  white-space: nowrap; /* Предотвращает перенос текста */
}

.word {
  background-color: #add8e6;
  top: 10px;
}

.phoneme {
  background-color: #ffff99;
  top: 120px;
}
