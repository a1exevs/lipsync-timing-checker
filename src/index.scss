@import url('https://fonts.googleapis.com/css2?family=Roboto:ital,wght@0,100;0,300;0,400;0,500;0,700;0,900;1,100;1,300;1,400;1,500;1,700;1,900&display=swap');

* {
  padding: 0;
  margin: 0;
  box-sizing: border-box;
  font-family: 'Roboto', sans-serif;
}

body {
  margin: 0 auto;
  font-family: 'Roboto', sans-serif;
  background: #000000;

  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5rem;
  color: #e8e6e3;
}

.App {
  background: rgba(24, 26, 27, 0.75);
  margin: 0 auto;
  max-width: 1920px;
  min-height: 100vh;
  padding: 0;
  &__Title {
    text-align: center;
  }
}

.Waveform {
  display: flex;
  height: 200px;
  overflow-x: auto;
  overflow-y: hidden;
  border: 1px solid #ccc;
}

.Controls {
  display: flex;
  align-items: center;
  gap: 16px;
  padding: 8px 16px;
}

.JSONDataContainer {
  position: relative;
  height: 100px;
  border: 1px solid #ccc;
  overflow-x: auto;
  overflow-y: hidden;
  scrollbar-color: #505354 #707374;
  scrollbar-width: thin;
  &_withoutScrollbar {
    scrollbar-width: none;
  }
}

.JSONData {
  position: relative;
  height: 100px;
}

.PositionIndicator {
  position: absolute;
  top: 0;
  height: 100%;
  width: 2px;
  background-color: red;
  pointer-events: none;
  z-index: 10;
}

.Word,
.Phoneme {
  color: black;
  position: absolute;
  padding: 5px 10px;
  border: 1px solid #000;
  border-radius: 4px;
  cursor: pointer;
  white-space: nowrap;
}

.Word {
  background-color: #add8e6;
}

.Phoneme {
  background-color: #ffff99;
}
