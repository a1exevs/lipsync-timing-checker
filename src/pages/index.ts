import HomePage from 'src/pages/home/ui/home';

export { HomePage };
