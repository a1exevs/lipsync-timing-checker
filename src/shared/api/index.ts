import { arrayToObject } from 'src/shared/api/helpers/arrays';
import { getExtensionByFile, getFileData } from 'src/shared/api/helpers/files';

export { arrayToObject, getFileData, getExtensionByFile };
