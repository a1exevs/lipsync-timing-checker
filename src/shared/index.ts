// ui
export {
  Button,
  FilePicker,
  IconButton,
  useConfirmationDialog,
  useInfoDialog,
  DialogResponse,
  useDialogActions,
  ConfirmationDialogConfirmButton,
  ConfirmationDialogCancelButton,
  DialogProvider,
} from 'src/shared/ui';

// api
export {
  arrayToObject,
  getExtensionByFile,
  getFileData,
  capitalize,
  capitalizeLabel,
  throttleTime,
} from 'src/shared/model';

// model
export type { KeyOfValue } from 'src/shared/model';
export {
  currentLang,
  DESKTOP_4K_SCREEN_WIDTH_PX,
  DESKTOP_SCREEN_WIDTH_PX,
  TABLET_SCREEN_WIDTH_PX,
  MOBILE_SCREEN_WIDTH_PX,
} from 'src/shared/model';
