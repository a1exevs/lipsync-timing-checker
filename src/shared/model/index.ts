// helpers
export { arrayToObject } from 'src/shared/model/helpers/arrays';
export { getExtensionByFile, getFileData } from 'src/shared/model/helpers/files';
export { throttleTime } from 'src/shared/model/helpers/intervals';
export { capitalize, capitalizeLabel } from 'src/shared/model/helpers/strings';

// lang
export { currentLang } from 'src/shared/model/lang/lang.helpers';

// types
export type { KeyOfValue } from 'src/shared/model/types/common';
