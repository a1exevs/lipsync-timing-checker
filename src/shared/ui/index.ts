import Button from 'src/shared/ui/button/button';
import Dialog from 'src/shared/ui/dialog/dialog';
import {
  CancelButton as ConfirmationDialogCancelButton,
  ConfirmButton as ConfirmationDialogConfirmButton,
} from 'src/shared/ui/dialog/dialog.compounds';
import DialogProvider from 'src/shared/ui/dialog/dialog.provider';
import { DialogResponse } from 'src/shared/ui/dialog/dialog.types';
import useConfirmationDialog from 'src/shared/ui/dialog/hooks/use-confirmation-dialog';
import useDialogActions from 'src/shared/ui/dialog/hooks/use-dialog-actions';
import useInfoDialog from 'src/shared/ui/dialog/hooks/use-info-dialog';
import FilePicker from 'src/shared/ui/file-picker/file-picker';
import IconButton from 'src/shared/ui/icon-button/icon-button';

export {
  Button,
  FilePicker,
  IconButton,
  Dialog,
  DialogProvider,
  useConfirmationDialog,
  useInfoDialog,
  DialogResponse,
  useDialogActions,
  ConfirmationDialogConfirmButton,
  ConfirmationDialogCancelButton,
};
