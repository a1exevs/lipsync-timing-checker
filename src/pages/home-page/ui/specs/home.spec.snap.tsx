import { render } from '@testing-library/react';
import React from 'react';

import { HomePage } from 'src/pages';
import { DialogProvider } from 'src/shared';

describe('HomePage', () => {
  beforeEach(() => {
    jest.spyOn(Storage.prototype, 'getItem').mockReturnValue('true');
    jest.spyOn(Storage.prototype, 'setItem').mockImplementation(() => {});
    jest.spyOn(Storage.prototype, 'removeItem').mockImplementation(() => {});
    jest.spyOn(Storage.prototype, 'clear').mockImplementation(() => {});
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('matches snapshot', () => {
    const { container } = render(
      <DialogProvider>
        <HomePage />
      </DialogProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
