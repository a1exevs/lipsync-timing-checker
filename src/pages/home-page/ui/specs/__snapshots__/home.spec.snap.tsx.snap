// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`HomePage matches snapshot 1`] = `
<div>
  <section
    class="flex flex-col"
  >
    <header
      class="grid grid-cols-[auto,1fr,auto] gap-4 items-center px-4 py-3 bg-gray-800 border-b border-gray-700 shadow-md shadow-black/30"
    >
      <svg
        aria-hidden="true"
        class="w-16 h-16"
      />
      <h1
        class="text-2xl font-extrabold tracking-tight"
      >
        Lipsync Timing Checker
      </h1>
      <button
        class="flex items-center justify-center rounded-full transition p-1 w-8 h-8 bg-gray-100 hover:bg-gray-200 text-gray-800 border border-gray-300 disabled:cursor-not-allowed disabled:opacity-50"
        title="Info"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-info"
          fill="none"
          height="18"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <path
            d="M12 16v-4"
          />
          <path
            d="M12 8h.01"
          />
        </svg>
      </button>
    </header>
    <section
      class="flex flex-col md:flex-row md:justify-between items-center my-4 gap-4 rounded-md bg-gray-800 p-4 shadow"
    >
      <div
        class="flex flex-col gap-3 w-full md:w-1/2"
      >
        <div
          class="flex flex-row gap-2 items-center"
        >
          <button
            class="inline-flex items-center justify-center gap-2 rounded-sm px-4 py-2 text-sm font-medium transition disabled:cursor-not-allowed disabled:opacity-50 bg-blue-600 hover:bg-blue-700 text-white w-1/2"
            type="button"
          >
            <span>
              <svg
                aria-hidden="true"
                class="lucide lucide-file-audio"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M17.5 22h.5a2 2 0 0 0 2-2V7l-5-5H6a2 2 0 0 0-2 2v3"
                />
                <path
                  d="M14 2v4a2 2 0 0 0 2 2h4"
                />
                <path
                  d="M2 19a2 2 0 1 1 4 0v1a2 2 0 1 1-4 0v-4a6 6 0 0 1 12 0v4a2 2 0 1 1-4 0v-1a2 2 0 1 1 4 0"
                />
              </svg>
            </span>
            <span>
              Load audio
            </span>
          </button>
          <input
            accept="audio/*"
            data-testid="file-input-audio"
            style="display: none;"
            type="file"
          />
        </div>
        <div
          class="flex flex-row gap-2 items-center"
        >
          <button
            class="inline-flex items-center justify-center gap-2 rounded-sm px-4 py-2 text-sm font-medium transition disabled:cursor-not-allowed disabled:opacity-50 bg-blue-600 hover:bg-blue-700 text-white w-1/2"
            disabled=""
            type="button"
          >
            <span>
              <svg
                aria-hidden="true"
                class="lucide lucide-file-json"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V7Z"
                />
                <path
                  d="M14 2v4a2 2 0 0 0 2 2h4"
                />
                <path
                  d="M10 12a1 1 0 0 0-1 1v1a1 1 0 0 1-1 1 1 1 0 0 1 1 1v1a1 1 0 0 0 1 1"
                />
                <path
                  d="M14 18a1 1 0 0 0 1-1v-1a1 1 0 0 1 1-1 1 1 0 0 1-1-1v-1a1 1 0 0 0-1-1"
                />
              </svg>
            </span>
            <span>
              Load JSON data
            </span>
          </button>
          <input
            accept="application/json"
            data-testid="file-input-json"
            style="display: none;"
            type="file"
          />
        </div>
      </div>
      <div
        class="flex flex-col gap-3 w-full md:w-auto"
      >
        <button
          class="inline-flex items-center justify-center gap-2 rounded-sm px-4 py-2 text-sm font-medium transition disabled:cursor-not-allowed disabled:opacity-50 bg-gray-100 hover:bg-gray-200 text-gray-800 border border-gray-300 w-full"
          type="button"
        >
          <span>
            <svg
              aria-hidden="true"
              class="lucide lucide-rocket"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4.5 16.5c-1.5 1.26-2 5-2 5s3.74-.5 5-2c.71-.84.7-2.13-.09-2.91a2.18 2.18 0 0 0-2.91-.09z"
              />
              <path
                d="m12 15-3-3a22 22 0 0 1 2-3.95A12.88 12.88 0 0 1 22 2c0 2.72-.78 7.5-6 11a22.35 22.35 0 0 1-4 2z"
              />
              <path
                d="M9 12H4s.55-3.03 2-4c1.62-1.08 5 0 5 0"
              />
              <path
                d="M12 15v5s3.03-.55 4-2c1.08-1.62 0-5 0-5"
              />
            </svg>
          </span>
          <span>
            Load example
          </span>
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 rounded-sm px-4 py-2 text-sm font-medium transition disabled:cursor-not-allowed disabled:opacity-50 bg-red-600 hover:bg-red-700 text-white w-full"
          disabled=""
          type="button"
        >
          <span>
            <svg
              aria-hidden="true"
              class="lucide lucide-download"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 15V3"
              />
              <path
                d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
              />
              <path
                d="m7 10 5 5 5-5"
              />
            </svg>
          </span>
          <span>
            Download JSON data
          </span>
        </button>
      </div>
    </section>
    <section
      class="flex items-center gap-6 my-4 rounded-md bg-gray-800 p-4 shadow"
    >
      <button
        class="flex items-center justify-center rounded-full transition p-2 w-10 h-10 bg-blue-600 hover:bg-blue-700 text-white disabled:cursor-not-allowed disabled:opacity-50"
        disabled=""
        title="Play"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-play"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="6 3 20 12 6 21 6 3"
          />
        </svg>
      </button>
      <div
        class="flex items-center gap-2"
      >
        <button
          class="flex items-center justify-center rounded-full transition p-2 w-10 h-10 bg-gray-100 hover:bg-gray-200 text-gray-800 border border-gray-300 disabled:cursor-not-allowed disabled:opacity-50"
          disabled=""
          title="Lock caret position"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-pin"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 17v5"
            />
            <path
              d="M9 10.76a2 2 0 0 1-1.11 1.79l-1.78.9A2 2 0 0 0 5 15.24V16a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1v-.76a2 2 0 0 0-1.11-1.79l-1.78-.9A2 2 0 0 1 15 10.76V7a1 1 0 0 1 1-1 2 2 0 0 0 0-4H8a2 2 0 0 0 0 4 1 1 0 0 1 1 1z"
            />
          </svg>
        </button>
        <button
          class="flex items-center justify-center rounded-full transition p-2 w-10 h-10 bg-gray-100 hover:bg-gray-200 text-gray-800 border border-gray-300 disabled:cursor-not-allowed disabled:opacity-50"
          disabled=""
          title="Enable audio while dragging"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-ear"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 8.5a6.5 6.5 0 1 1 13 0c0 6-6 6-6 10a3.5 3.5 0 1 1-7 0"
            />
            <path
              d="M15 8.5a2.5 2.5 0 0 0-5 0v1a2 2 0 1 1 0 4"
            />
          </svg>
        </button>
      </div>
    </section>
    <section>
      <section
        class="flex overflow-x-auto overflow-y-hidden border border-gray-300"
        style="scrollbar-width: none;"
      >
        <canvas
          height="50"
          style="display: block; width: 0px; height: 50px;"
          width="0"
        />
      </section>
      <section
        class="flex h-[200px] overflow-x-auto overflow-y-hidden border border-gray-300"
        style="scrollbar-width: none;"
      />
      <section>
        <div
          class="relative h-[100px] border border-gray-300 overflow-x-auto overflow-y-hidden"
          style="scrollbar-color: #505354 #707374; scrollbar-width: thin;"
        >
          <div
            class="relative h-[100px]"
          >
            <div
              class="absolute top-0 h-full w-[2px] bg-red-500 pointer-events-none z-10"
            />
          </div>
        </div>
      </section>
    </section>
  </section>
</div>
`;
